EIPS/eip-memory-paging.md

```diff
@@ -0,0 +1,276 @@
+-            # read() and not released. Callers of read() will never try to write to the
+-            # buffer so we're not missing anything by making a new buffer and forgetting
+-            # about the old one. We're keeping too much memory around but this is still
+-            # a net savings over having read() return a new bytes() object every time.
+-            self._bytes = self._bytes + bytearray(size_to_extend)
+-
+-    def __len__(self) -> int:
+-        return len(self._bytes)
++        if start_position + size >= 2**32:
++            raise PyEVMError("Non 32-bit address")
+
+-    def write(self, start_position: int, size: int, value: bytes) -> None:
+-        if size:
+-            validate_uint256(start_position)
+-            validate_uint256(size)
+-            validate_is_bytes(value)
+-            validate_length(value, length=size)
+-            validate_lte(start_position + size, maximum=len(self))
++        validate_uint256(start_position)
++        validate_uint256(size)
++        validate_is_bytes(value)
++        validate_length(value, length=size)
++
++        end_position = start_position + size
+
+-            self._bytes[start_position : start_position + len(value)] = value
++        self.memview[start_position : end_position] = value
+
++    # unused
+     def read(self, start_position: int, size: int) -> memoryview:
+-        return memoryview(self._bytes)[start_position : start_position + size]
++        return memoryview(self.memview)[start_position : start_position + size]
+
+     def read_bytes(self, start_position: int, size: int) -> bytes:
+-        return bytes(self._bytes[start_position : start_position + size])
++        return bytes(self.memview[start_position : start_position + size])
+
+     def copy(self, destination: int, source: int, length: int) -> None:
+         if length == 0:
+@@ -69,5 +68,5 @@ class Memory(MemoryAPI):
+         validate_uint256(length)
+         validate_lte(max(destination, source) + length, maximum=len(self))
+
+-        buf = memoryview(self._bytes)
++        buf = memoryview(self.memview)
+         buf[destination : destination + length] = buf[source : source + length]
+```
+
+## Security Considerations
+
+There are two primary security considerations regarding this EIP.
+
+One, does it break existing contracts? That is, do existing contracts depend on memory being restricted?
+
+Outside of gas costing, existing contracts may simply execute to completion rather than running out of gas.
+
+Two, does this enable memory-based DoS attacks against clients?
+
+This requires a maximum-memory usage analysis. Based on a gas limit of 30mm gas, and page allocation cost of 100, the maximum amount of memory that can be physically allocated on a client is 1.1GB. Note that this analysis holds irrespective of the shape of the call stack. While this is substantially more than what can be allocated today, it is still insignificant compared to memory limits on modern computers.
```

Comment:
This may enable a DoS attack for concurrent eth_call's - if a server with 32GB memory is processing 30 eth_call's, each of which consumes 1.1GB memory, the total memory cost will be 33GB, resulting in a potential node crash.

Reply:
perhaps a global limit is called for, then